sequoia/methods/packnet_method.py

```diff
@@ -0,0 +1,277 @@
+
+    def mask_remaining_params(self, model):
+        """
+        Create mask for remaining parameters
+        """
+        mask_idx = 0
+        mask = []
+        for mod in model.children():
+            if isinstance(mod, nn.Conv2d) or isinstance(mod, nn.Linear):
+                for name, param_layer in mod.named_parameters():
+                    if 'bias' not in name:
+
+                        # Get mask of weights from previous tasks
+                        prev_mask = torch.zeros(param_layer.size(), dtype=torch.bool, requires_grad=False)
+                        for task in self.masks:
+                            prev_mask |= task[mask_idx]
+
+                        # Create mask of remaining parameters
+                        layer_mask = ~prev_mask
+                        mask.append(layer_mask)
+
+                        mask_idx += 1
+        self.masks.append(mask)
+
+    def total_epochs(self):
+        return self.epoch_split[0] + self.epoch_split[1]
+
+    def config_instructions(self):
+        """
+        Create pruning instructions for this task split
+        :return: None
+        """
+        assert self.n_tasks is not None
+
+        if not isinstance(self.prune_instructions, list):  # if a float is passed in
+            assert 0 < self.prune_instructions < 1
+            self.prune_instructions = [self.prune_instructions] * (self.n_tasks - 1)
+        assert len(self.prune_instructions) == self.n_tasks - 1, "Must give prune instructions for each task"
+
+    def save_final_state(self, model, PATH='model_weights.pth'):
+        """
+        Save the final weights of the model after training
+        :param model: pl_module
+        :param PATH: The path to weights file
+        """
+        self.PATH = PATH
+        torch.save(model.state_dict(), PATH)
+
+    def load_final_state(self, model):
+        """
+        Load the final state of the model
+        """
+        model.load_state_dict(torch.load(self.PATH))
+
+    def on_init_end(self, trainer):
+        self.mode = 'train'
+
+    def on_after_backward(self, trainer, pl_module):
+
+        if self.mode == 'train':
+            self.training_mask(pl_module)
+
+        elif self.mode == 'fine_tune':
+            self.fine_tune_mask(pl_module)
+
+    def on_epoch_end(self, trainer, pl_module):
+
+        if pl_module.current_epoch == self.epoch_split[0] - 1:  # Train epochs completed
+            self.mode = 'fine_tune'
+            if self.current_task == self.n_tasks - 1:
+                self.mask_remaining_params(pl_module)
+            else:
+                self.prune(
+                    model=pl_module,
+                    prune_quantile=self.prune_instructions[self.current_task])
+
+        elif pl_module.current_epoch == self.total_epochs() - 1:  # Train and fine tune epochs completed
+            self.save_final_state(pl_module)
+            self.mode = 'train'
+
+
+class PackNetMethod(BaseMethod, target_setting=TaskIncrementalSLSetting):
```

Comment:
You're currently inheriting from `BaseMethod`, while you most likely meant to inherit from `Method` (the abstract base class for all methods). The naming isn't super intuitive.

It would be nice if this PackNet method were to actually inherit from `BaseMethod`, so it could be applied in RL and SL (via the BaseModel), by just changing what callbacks get created. You'd have to basically add a parameter to the BaseMethod constructor, and change the callbacks that are passed to the Trainer. I can help you with that.